libs/state/src/lib/transformation-helpers/array/upsert.ts

```diff
@@ -0,0 +1,48 @@
+import { isDefined } from '../../core/utils/guards';
+import { valuesComparer } from '../_internals/valuesComparer.util';
+import { ComparableData } from '../interfaces/comparable-data-type';
+
+export function upsert<T extends object>(
+  source: T[],
+  update: T | T[] | Partial<T>[] | Partial<T>,
+  compare?: ComparableData<T>
+): T[] {
+  const updatesAsArray = update
+                         ? Array.isArray(update)
+                           ? update
+                           : [update]
+                         : [];
+
+  const sourceDefined = isDefined(source);
+  const sourceIsArray = Array.isArray(source);
+  const invalidInput = !sourceIsArray && !isDefined(update);
+
+  if (invalidInput) {
+    return source;
+  }
+
+  if (!sourceDefined || !source.length || !sourceIsArray) {
+    return [...updatesAsArray] as T[];
+  }
+
+  const inserts: T[] = [];
+  const updates: Record<number, Partial<T>> = {};
+  updatesAsArray.forEach(item => {
+    const match = source.findIndex(sourceItem => valuesComparer(item as T, sourceItem, compare));
```

Comment:
Do we care about IE 11? If yes, `findIndex` is not supported in this browser-like application.

Reply:
is this really a thing? this library gets transpiled, doesn't it?

Reply:
In ng10+ yes, in previous versions, there might be some issues. 

But the proper fix for previous ng versions and old browsers will be to downgrade ng-packagr back to version 9.X.X, so it will produce fesm5 and esm5. Not sure if we should do that :)

Reply:
i don't know if we should take care of browsers which are deprecated especially when keeping in mind that we are offering cutting edge technology with rx-angular. @BioPhoton what's your opinion about this?

Reply:
> Do we care about IE 11? If yes, `findIndex` is not supported in this browser-like application.

We should discuss that one soon! #303

Reply:
@BioPhoton imo IE11 should not get supported by a cutting edge library. And didn't angular support default IE11 support as well?

Reply:
but if it's just about this line, i wont complain if we change it :+1: